### PR TITLE
CLN: rename reduce-->do_reduce

### PR DESCRIPTION
**Pull request by jbrockmendel**
The current name makes it difficult to grep for.

Assorted cleanups; added type annotations in a few places.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -224,8 +213,8 @@ class TimelikeOps:
 
             .. versionadded:: 0.24.0
 
-        nonexistent : 'shift_forward', 'shift_backward', 'NaT', timedelta, \
-default 'raise'
+        nonexistent : 'shift_forward', 'shift_backward', 'NaT', timedelta,
+            default 'raise'
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimelike.py]: this is for proper rendering with numpydoc

File: pandas/core/arrays/datetimes.py [hunk 2]
```diff
@@ -328,7 +328,6 @@ class DatetimeArray(dtl.DatetimeLikeArrayMixin, dtl.TimelikeOps, dtl.DatelikeOps
     # -----------------------------------------------------------------
     # Constructors
 
-    _attributes = ["freq", "tz"]
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimes.py]: Why are you removing those attributes? (I suppose they are thus not used) It's a left-over from when it shared code with the Index?

File: pandas/_libs/reduction.pyx [hunk 3]
```diff
@@ -628,7 +628,7 @@ cdef class BlockSlider:
             arr.shape[1] = 0
 
 
-def reduce(arr, f, axis=0, dummy=None, labels=None):
+def do_reduce(arr, f, axis=0, dummy=None, labels=None):
```
Comment (jreback) [inline pandas/_libs/reduction.pyx]: I don't particular like this name, we don't have anywhere we call things do_*, how about
compute_reduction

**Review by jreback — CHANGES_REQUESTED** [review]
I would also change _reduce in ExtensionArray to match

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: darn, will revert.

File: pandas/core/arrays/datetimes.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: exactly, leftover from the inheritance days

File: pandas/_libs/reduction.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/reduction.pyx]: sure

Comment (jbrockmendel) [reply pandas/_libs/reduction.pyx]: (and its such an easy change to make because I can grep for it!)

**Comment by jreback** [comment]
> I would also change _reduce in ExtensionArray to match

need this too

**Comment by jorisvandenbossche** [comment]
The `_reduce` is part of the EA interface, so we cannot just rename that.

Besides, that seems like to defeat Brock's purpose to be able to search for a specific function (and thus not having multiple functions or methods with the same name that do a different thing)

**Comment by jbrockmendel** [comment]
We have _reduce defined in a bunch of places, renaming it should be a separate discussion

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**